doc/architecture/decisions/0004-generate-domain-events-when-ap-data-is-changed.md

```diff
@@ -0,0 +1,128 @@
+
+## Decision
+
+In order to support those needs we are adopting the HMPPS pattern of publishing Domain
+Events which represent changes to the data which our service "masters". Following the
+guidance from the HMPPS Technical Architect community:
+
+- [HMPPS ADR - Domain Event Schema](https://dsdmoj.atlassian.net/wiki/spaces/NDSS/pages/3732570166/ADR+-+Domain+Event+Schema)
+- [HMPPS ADR - Publishing Microservice Events](https://dsdmoj.atlassian.net/wiki/spaces/NDSS/pages/2811691200/ADR+-+Publishing+Microservice+Events)
+
+we will implement the following process:
+
+### 1. Our service emits a Domain Event message
+
+
+A lightweight message is broadcast to a system-wide channel announcing the type of event
+and providing a link to full event information (`detailUrl`) e.g.
+
+```json
+{
+  "eventType": "approved-premises.person.arrived",
+  "version": 1,
+  "description": "A person has arrived at an Approved Premises",
+  "detailUrl": "https://approved-premses/events/person-arrived/68df9f6c-3fcb-4ec6-8fcf-96551cd9b080",
+  "occurredAt": "2022-12-04T10:42:43+00:00",
+  "additionalInformation": {
+    "applicationId": "68df9f6c-3fcb-4ec6-8fcf-96551cd9b080"
+  }
+  "personReference": {
+    "identifiers": [
+      {"type": "CRN", "value":"X08769"},
+      {"type": "NOMS","value":"A1234CR"}
+    ]
+  }
+}
+```
+
+This "pub-sub" mechanism will use AWS SNS subscriptions and SQS queues in the HMPPS-wide
+`domain-events` namespace.
+
+### 2. Delius integration obtains full details
+
+The interested party, in this case the "Delius integration" which intends to use the
+domain event in order to update Delius, follows the `detailsUrl` provided to obtain the
+full details of the event from our "events" API. The representation describes who arrived
+at which AP, and when, e.g:
+
+```json
+{
+  "id": "364145f9-0af8-488e-9901-b4c46cd9ba37",
+  "timestamp": "2022-11-30T14:53:44",
+  "eventType": "approved-premises.person.arrived",
+  "eventDetails": {
+    "personReference": {
+      "crn": "C123456",
+      "noms": "A1234ZX"
+    },
+    "bookingId": "14c80733-4b6d-4f35-b724-66955aac320c",
+    "premises": {
+      "id": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
+      "name": "Hope House",
+      "apCode": "NEHOPE1",
+      "legacyApCode": "Q057",
+      "probationArea": {
+        "code": "N54",
+        "name": "North East Region"
+      }
+    },
+    "keyWorker": {
+      "staffCode": "N54A999",
+      "staffIdentifier": 1501234567,
+      "forenames": "John",
+      "surname": "Smith",
+      "username": "JohnSmithNPS"
+    },
+    "arrivedAt": "2022-11-30T14:50:00",
+    "expectedDepartureOn": "2023-02-28",
+    "notes": "Arrived a day late due to rail strike. Informed in advance by COM."
+  }
+}
+```
+
+### 3. Delius integration updates NDelius
+
+This information is used to make changes to Delius, e.g. updating the person's address,
+creating a "contact" records within NDelius etc
+
+## Consequences
```

Comment:
Should we talk about the need to spin up new infrastructure for this?

Reply:
It's also worth pointing out this will have some impact on how quickly we deliver other features. I know we've factored this in, but worth documenting.

Reply:
I don't think we need to spin up any infrastructure for this. My understanding is that the event messages will be published to an existing AWS SNS queue called `hmpps-domain-events` and the full Domain Event representations will be exposed via our existing API, through an `/events/` namespace.

I totally agree that doing this will require effort have an impact on our overall capacity but I think that impact is reduced as this is a largely completely separate work-stream and doesn't complicate delivering apply, assess, match etc

Reply:
👍